app/guided.py

```diff
@@ -0,0 +1,388 @@
+        else:
+            new_distance = float(effective_score) / self.half_life_penalty
+        
+        return new_distance
+
+    async def _get_goal_links(self, absolute_distance_table, agent=None):
+        """
+        Produces all links for abilities that are present in the absolute distance table.
+        """
+        links = await self.planning_svc.get_links(self.operation, agent=agent)
+        return [link for link in links if link.ability.ability_id in absolute_distance_table]
+
+    async def _task_agents(self, links, link_distance_table, abilities, agent=None):
+        """
+        Chooses next actions for agent provided. If no agent is provided, then the method will choose and return
+        next actions for all agents in the operation.
+        """
+        tasked_link_ids = []
+        if agent:
+            agent_links = [link for link in links if link.paw == agent.paw]
+            if len(agent_links):
+                tasked_link = await self._task_best_action(agent_links, link_distance_table, abilities)
+                tasked_link_ids.append(tasked_link.id)
+        else:
+            for operation_agent in self.operation.agents:
+                agent_links = [link for link in links if link.paw == operation_agent.paw]
+                if not len(agent_links):
+                    continue
+                tasked_link = await self._task_best_action(agent_links, link_distance_table, abilities)
+                tasked_link_ids.append(tasked_link.id)
+
+        return tasked_link_ids
+
+    async def _task_best_action(self, agent_links, link_distance_table, abilities):
+        """
+        Produces the best available link based on the current distance table. The method also applies links
+        that support the chosen link, but that do not actively work towards the goal.
+        """
+        async def _heapsort(links):
+            sorted_links = []
+            counter = itertools.count()
+            for _, (dist, link) in links.items():
+                entry = [-dist, -link.score, next(counter), link]
+                heapq.heappush(sorted_links, entry)
+            return [heapq.heappop(sorted_links)[3] for _ in range(len(sorted_links))]
+
+        weighted_scores = {link.id: ((link_distance_table[link.ability.ability_id] * self.goal_weight + 
+                                      link.score * self.fact_score_weight), link) for link in agent_links}
+        sorted_links = await _heapsort(weighted_scores)
+        link_to_execute = sorted_links[0]
+
+        supporting_links = await self._get_supporting_links(link_to_execute, abilities)
+        for supporting_link in supporting_links:
+            await self.operation.apply(supporting_link)
+        await self.operation.apply(link_to_execute)
+        self.last_action = link_to_execute.ability
+        return link_to_execute
+
+    async def _get_supporting_links(self, link, abilities):
```

Comment:
This method is a quick fix that ideally should be resolved by improvements to abilities at some point in the future. It's here to capture orphaned abilities that do not have any output facts. Within the dependency graph, they would look like single nodes that are not directly connected to any other components. It is still possible given the way that some abilities are implemented that some of these orphaned abilities are expected to be run before others even through there is not a fact dependency between them. This method gets around this by selecting all links with no output facts that are earlier in the atomic ordering that the currently chosen link. This guarantees that all possible dependent actions are taken prior to the chosen action.

Reply:
This is a great fix for the current set of adversaries and abilities.  You could consider expanding the docstring with some of the background from this comment.  With context, the logic looks great,